src/overlays/actors/ovl_En_Rat/z_en_rat.c

```diff
@@ -28,123 +47,913 @@ const ActorInit En_Rat_InitVars = {
+    Matrix_RotateZS(this->actor.shape.rot.z, MTXMODE_APPLY);
+
+    Matrix_MultVecZ(1.0f, &this->axisForwards);
+    Matrix_MultVecY(1.0f, &this->axisUp);
+    Matrix_MultVecX(1.0f, &this->axisLeft);
+}
+
+/**
+ * Returns true if floorPoly is valid for the Real Bombchu to move on, false otherwise.
+ */
+s32 EnRat_UpdateFloorPoly(EnRat* this, CollisionPoly* floorPoly, PlayState* play) {
+    Vec3f normal;
+    Vec3f vec;
+    f32 angle;
+    f32 magnitude;
+    f32 normDotUp;
+
+    this->actor.floorPoly = floorPoly;
+
+    if (floorPoly != NULL) {
+        normal.x = COLPOLY_GET_NORMAL(floorPoly->normal.x);
+        normal.y = COLPOLY_GET_NORMAL(floorPoly->normal.y);
+        normal.z = COLPOLY_GET_NORMAL(floorPoly->normal.z);
+    } else {
+        normal.x = 0.0f;
+        normal.z = 0.0f;
+        normal.y = 1.0f;
+    }
+
+    normDotUp = DOTXYZ(normal, this->axisUp);
+
+    if (fabsf(normDotUp) >= 0.999f) {
+        return false;
+    }
+
+    angle = func_80086C48(normDotUp);
+    if (angle < 0.001f) {
+        return false;
+    }
+
+    Math3D_CrossProduct(&this->axisUp, &normal, &vec);
+
+    magnitude = Math3D_Vec3fMagnitude(&vec);
+
+    if (magnitude < 0.001f) {
+        EnRat_Explode(this, play);
+        return false;
+    }
+
+    Math_Vec3f_Scale(&vec, 1.0f / magnitude);
+    Matrix_RotateAxisF(angle, &vec, MTXMODE_NEW);
+    Matrix_MultVec3f(&this->axisLeft, &vec);
+    Math_Vec3f_Copy(&this->axisLeft, &vec);
+    Math3D_CrossProduct(&this->axisLeft, &normal, &this->axisForwards);
+
+    magnitude = Math3D_Vec3fMagnitude(&this->axisForwards);
+    if (magnitude < 0.001f) {
+        EnRat_Explode(this, play);
+        return false;
+    }
+
+    Math_Vec3f_Scale(&this->axisForwards, 1.0f / magnitude);
+    Math_Vec3f_Copy(&this->axisUp, &normal);
+    return true;
+}
+
+void EnRat_UpdateRotation(EnRat* this) {
+    MtxF mf;
+
+    mf.xx = this->axisLeft.x;
+    mf.yx = this->axisLeft.y;
+    mf.zx = this->axisLeft.z;
+
+    mf.xy = this->axisUp.x;
+    mf.yy = this->axisUp.y;
+    mf.zy = this->axisUp.z;
+
+    mf.xz = this->axisForwards.x;
+    mf.yz = this->axisForwards.y;
+    mf.zz = this->axisForwards.z;
+
+    Matrix_MtxFToYXZRot(&mf, &this->actor.world.rot, false);
+    this->actor.world.rot.x = -this->actor.world.rot.x;
+}
+
+/**
+ * Chooses to either look at the player (if the Real Bombchu has spotted the player or
+ * is chasing after them) or at some other semi-random point (if the Bombchu is idle).
+ */
+void EnRat_ChooseDirection(EnRat* this) {
+    Vec3f sp74;
+    s16 angle;
+
+    if (this->actionFunc != EnRat_Idle) {
+        angle = this->actor.yawTowardsPlayer - this->actor.shape.rot.y;
+        if (this->axisUp.y < -0.25f) {
+            angle -= 0x8000;
+        }
+    } else {
+        if (Actor_DistanceToPoint(&this->actor, &this->actor.home.pos) > 50.0f) {
+            Vec3f sp64;
+            Vec3f sp58;
+            Vec3f sp4C;
+            Vec3f sp40;
```

Comment:
Not quite sure what to name these/`sp74` myself.

Reply:
Yeah, I'm sure there's just some math thing I'm unfamiliar with here. I wish I could name them, but I'm not going to pretend like I know what these are.

Reply:
here may be a better name for axisUp, but it seems to be like a cameras up vector. It's the norm of the floor poly this is standing on.

So first a new matrix is created matrix is rotated based on the negative angles on the actor.
Then axisUp is multiplied into the current matrix and placed in sp40. 
Then the current position has the axis forwarded to it, which would be the tangent of the collision poly. This result is placed in sp64.
Then sp64 is multiplied into the current matrix and placed in sp4c.
Then the current home position is multiplied into the current matrix and placed in sp64 (seems reused)
Then the current world position is multiplied into the current matrix and placed into sp58.
Then it looks for the yaw between sp58 and sp64 then subtract the yaw between sp58 and sp4c from the original yaw and sets that to angle.
The angle is then randomized a little.

The angle is the output of this calculation.

Does this help you name it at all?

Reply:
If I'm being honest, not particularly, no. I don't really understand what the matrix at the start is even supposed to represent.

Reply:
This can be added later. Can you put my comments in at least?

Reply:
As we talked about in Discord, I managed to sort of figure out what's going on here with debug prints. Feel free to critique the names, however, since I'm not going to claim I have a perfect understanding of this or anything.